src/test/scala/io/iohk/ethereum/blockchain/sync/regular/NewRegularSyncSpec.scala

```diff
@@ -0,0 +1,633 @@
+              DuplicateBlock
+            } else {
+              if (importedBlocks.isEmpty || Block.isParentOf(bestBlock, block)) {
+                importedBlocks.add(block)
+                BlockImportedToTop(List(BlockData(block, Nil, block.header.difficulty)))
+              } else if (Block.number(block) > Block.number(bestBlock)) {
+                importedBlocks.add(block)
+                BlockEnqueued
+              } else {
+                BlockImportFailed("foo")
+              }
+            }
+
+            Future.successful(result)
+          }
+
+          override def resolveBranch(headers: scala.Seq[BlockHeader]): BranchResolutionResult = {
+            val importedHashes = importedBlocks.map(Block.hash).toSet
+
+            if (importedBlocks.isEmpty || (importedHashes.contains(headers.head.parentHash) && headers.last.number > Block
+                .number(bestBlock))) {
+              NewBetterBranch(Nil)
+            } else {
+              UnknownBranch
+            }
+          }
+        }
+      }
+
+      "go back to earlier block in order to find a common parent with new branch" in new Fixture(testSystem)
+      with FakeLedger {
+        implicit val ec: ExecutionContext = system.dispatcher
+        override lazy val syncConfig: SyncConfig =
+          defaultSyncConfig.copy(blockHeadersPerRequest = 5, blockBodiesPerRequest = 5, blocksBatchSize = 5)
+        override lazy val ledger: TestLedgerImpl = new FakeLedgerImpl()
+
+        val commonPart: List[Block] = testBlocks.take(syncConfig.blocksBatchSize)
+        val alternativeBranch: List[Block] = getBlocks(syncConfig.blocksBatchSize * 2, commonPart.last)
+        val alternativeBlocks: List[Block] = commonPart ++ alternativeBranch
+
+        class BranchResolutionAutoPilot(didResponseWithNewBranch: Boolean, blocks: List[Block])
+            extends PeersClientAutoPilot(blocks) {
+          override def overrides(sender: ActorRef): PartialFunction[Any, Option[AutoPilot]] = {
+            case PeersClient.Request(GetBlockHeaders(Left(nr), maxHeaders, _, _), _, _)
+                if nr >= alternativeBranch.numberAtUnsafe(syncConfig.blocksBatchSize) && !didResponseWithNewBranch =>
+              val responseHeaders = alternativeBranch.headers.filter(_.number >= nr).take(maxHeaders.toInt)
+              sender ! PeersClient.Response(defaultPeer, BlockHeaders(responseHeaders))
+              Some(new BranchResolutionAutoPilot(true, alternativeBlocks))
+          }
+        }
+
+        peersClient.setAutoPilot(new BranchResolutionAutoPilot(didResponseWithNewBranch = false, testBlocks))
+
+        Await.result(ledger.importBlock(genesis), remainingOrDefault)
+
+        regularSync ! RegularSync.Start
+
+        peerEventBus.expectMsgClass(classOf[Subscribe])
+        peerEventBus.reply(MessageFromPeer(NewBlock(testBlocks.last, Block.number(testBlocks.last)), defaultPeer.id))
+
+        awaitCond(ledger.bestBlock == alternativeBlocks.last, 10 minutes)
```

Comment:
isn't 10minutes too long?

Reply:
It is. I forgot to change that value after debugging.